plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1371,6 +1378,141 @@ \subsubsection{Comparison of the typed CK machine and the untyped CEK machine}
+  \item Converting \texttt{32768} to binary: \\
+    \verb|32768| $\rightarrow$ \verb|0b1000000000000000|
+  \item Split into 7 bit chunks: \\
+    \verb|0b1000000000000000| $\rightarrow$ \verb|0000010 0000000 0000000|
+  \item Reorder chunks (least significant chunk first): \\
+    \verb|0000010 0000000 0000000| $\rightarrow$ \verb|0000000 0000000 0000010|
+  \item Add list constructor tags: \\
+    \verb|0000000 0000000 0000010| $\rightarrow$ \verb|10000000 10000000 00000010|
+\end{enumerate}
+
+\noindent Encoding primtive types:
+
+\begin{itemize}
+  \item Unsigned \texttt{Integer}s and \texttt{ByteString}s are encoded using the
+    previously introduced encoding for variable length data types.
+  \item Signed \texttt{Integer}s are first converted to an unsigned value using the
+    \texttt{ZigZag} encoding, then they are encoded as variable length data types.
+  \item \texttt{String}s are encoded as lists of characters and use the \texttt{UTF-8}
+    encoding.
+  \item \texttt{Unit} and data structures which have only one constructor, are removed
+    from the serialised data.
+  \item Variable names are encoded using DeBruijn indices, which are unsigned, arbitrary
+    length integers.
+\end{itemize}
+
+Encoded values are aligned to byte/word boundary using a meaningless sequence of `0` bits
+terminated with a `1` bit.
+
+\subsection{Constants}
+
+Constants are encoded as a combination of the kind of value that is serialized and the
+value itself. Constants use 4 bits to encode the constructors, so they allow for a
+maximum of 16 constructors, of which 6 are currently used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|l|c|l|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Encoding} \\
+  \hline
+  ByteString & 1 & Variable length \rule{0mm}{4mm} \\[\sep]
+  Integer & 0 & ZigZag + Variable length \\[\sep]
+  String & 2 & UTF-8 \\[\sep]
+  Char & 3 & UTF-8 \\[\sep]
+  Unit & 4 & Empty \\[\sep]
+  Bool & 5 & `1` is True, `0` is False \\[\sep]
+  \hline
+\end{tabular}
+\captionof{figure}{Constant kinds}
+\label{fig:serialisation-constants}
+\end{minipage}
+
+\subsection{Untyped terms}
+
+Terms are encoded using 4 bits, which allows a total of 16 terms, of which 8 are
+currently used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|l|c|l|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Arguments} \\
+  \hline
+  Variable & 0 & name \rule{0mm}{4mm} \\[\sep]
+  Delay & 1 & term \\[\sep]
+  Lambda abstraction & 2 & name, term \\[\sep]
+  Application & 3 & term, term \\[\sep]
+  Constant & 4 & constant \\[\sep]
+  Force & 5 & term \\[\sep]
+  Error & 6 & term \\[\sep]
+  Builtin & 7 & builtin \\[\sep]
+  \hline
+\end{tabular}
+\captionof{figure}{Untyped terms}
+\label{fig:serialisation-terms}
+\end{minipage}
+
+\vspace{1cm}
+
+\noindent Builtins use 5 bits for their definition, allowing for a maximum of 32 builtin
+functions of which 25 are used.
+
```

Comment:
Having maximum of 32 builtin functions seems like a really, really bad idea. When we add bit ops for `Integer`s, `String` ops, not even saying about other crypto primitives, – and we are way over 32.

Reply:
This document is in sync with the code that we have right now, but realistically these values will change a lot as we get closer to the release. I am not sure if this is the best place to decide the number of builtins we will support.

My rule of thumb is to use a size that can accommodate we have right now.

Reply:
However, I don't really have a problem with increasing the number of bits at this point either. @nau do you have any suggestion?

Reply:
I'm just worrying that these sizes could be hard to change afterwards. It's good to know those are not set in stone yet. I'd discuss and establish an amount of _reserved_ builtins we should remain for the future, and state it explicitly. Like, we'd like to have, say, at least 32 reserved builtins or something.

Reply:
Increasing this limit would require changing the binary format of scripts. However, this is not so bad, because *most* changes to the language will require a "new" language, and thus it's relatively easy to say it has a new format.

It's *nice* to keep the format the same as much as possible, but it's not critical. (This reasoning should get written down somewhere...)

Reply:
That said... I'm sure flat has a way to encode arbitrary-length integers. It might be worth seeing how much space we use if we encode the builtin tags as arbitrary-length integers instead. If it's cheap we might as well do it, I guess.

Reply:
It uses 1 byte to encode a 7bit value, and it uses minimum 1 byte.

Reply:
Having thought about this some more, I'm wondering whether we should allow a variable-length encoding for built-in types and functions, like the one that you can derive automatically using `Generic`.  The automatically-generated encoding uses a prefix code (I think it's actually a truncated binary encoding, a Huffman encoding with uniform weights: see https://en.wikipedia.org/wiki/Truncated_binary_encoding).  Since it's a prefix code, the decoder can tell when it's got to the end of an encoding without having to know the length in advance.  

The reason I'm suggesting this is that the plan is to allow users to implement their own builtin types and functions, and (a) there's no reason why these should bear any relation to the builtins we have here, and (b) there may be more types/functions than you can fit into a predetermined number of bits.  With `flat`'s generic encoding we could just say that when you implement your builtins you use that specific encoding (perhaps implemented by hand if you're not using Haskell) and there would be no necessity to fix the size of the encoding in advance.  We should probably postpone that until another PR though.  At some point we're going to have to specify the mechanism for extensible builtins (SCP_1058) and that might be a good time to think about this again (ie, put it off for as long as possible).

By the way, is there any guarantee that `Generic` will always do the same thing?  Is there any danger that if you recompile your program in 10 years, it'll process constructors in a different way or something?   If that's a possibility it would mean that we definitely shouldn't use generically-generated encodings.